federation/pkg/kubefed/init/init.go

```diff
@@ -289,32 +312,50 @@ func createNamespace(clientset *client.Clientset, namespace string, dryRun bool)
 	return clientset.Core().Namespaces().Create(ns)
 }
 
-func createService(clientset *client.Clientset, namespace, svcName string, dryRun bool) (*api.Service, error) {
-	svc := &api.Service{
+func createService(clientset *client.Clientset, namespace, svcName, apiServiceType, apiServerAdvertiseAddress string, dryRun bool) (svc *api.Service, ips []string, hostnames []string, err error) {
+	serviceSpec := api.ServiceSpec{
+		Selector: apiserverSvcSelector,
+		Ports: []api.ServicePort{
+			{
+				Name:     "https",
+				Protocol: "TCP",
+				Port:     443,
+			},
+		},
+	}
+
+	if apiServiceType == "loadbalancer" {
+		serviceSpec.Type = api.ServiceTypeLoadBalancer
+		serviceSpec.Ports[0].TargetPort = intstr.FromInt(443)
+	} else {
+		serviceSpec.Type = api.ServiceTypeNodePort
+		serviceSpec.Ports[0].NodePort = int32(federationApiserverNodeport)
+	}
+
+	svc = &api.Service{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      svcName,
 			Namespace: namespace,
 			Labels:    componentLabel,
 		},
-		Spec: api.ServiceSpec{
-			Type:     api.ServiceTypeLoadBalancer,
-			Selector: apiserverSvcSelector,
-			Ports: []api.ServicePort{
-				{
-					Name:       "https",
-					Protocol:   "TCP",
-					Port:       443,
-					TargetPort: intstr.FromInt(443),
-				},
-			},
-		},
+		Spec: serviceSpec,
 	}
 
 	if dryRun {
-		return svc, nil
+		return svc, ips, hostnames, nil
+	}
+
+	if apiServiceType == serviceTypeLoadBalancer {
+		ips, hostnames, err = waitForLoadBalancerAddress(clientset, svc, dryRun)
```

Comment:
I felt it is sensible to move logically related code into common place instead of putting conditionals in main functions. and code-wise, i feel its the same, because anyway we need to pass apiServiceType to createService func.